src/main/java/org/ibp/api/java/impl/middleware/common/validator/MeansImportRequestValidator.java

```diff
@@ -0,0 +1,201 @@
+package org.ibp.api.java.impl.middleware.common.validator;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.apache.commons.collections.map.CaseInsensitiveMap;
+import org.apache.commons.lang.StringUtils;
+import org.generationcp.middleware.domain.dms.DataSet;
+import org.generationcp.middleware.domain.dms.Study;
+import org.generationcp.middleware.domain.ontology.Variable;
+import org.generationcp.middleware.domain.ontology.VariableType;
+import org.generationcp.middleware.enumeration.DatasetTypeEnum;
+import org.generationcp.middleware.manager.api.StudyDataManager;
+import org.generationcp.middleware.manager.ontology.api.OntologyVariableDataManager;
+import org.generationcp.middleware.manager.ontology.daoElements.VariableFilter;
+import org.generationcp.middleware.pojos.dms.StockModel;
+import org.generationcp.middleware.service.impl.analysis.MeansImportRequest;
+import org.ibp.api.exception.ApiRequestValidationException;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+import org.springframework.validation.BindingResult;
+import org.springframework.validation.MapBindingResult;
+import org.thymeleaf.util.MapUtils;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Component
+public class MeansImportRequestValidator {
+
+	@Autowired
+	private StudyDataManager studyDataManager;
+
+	@Autowired
+	private OntologyVariableDataManager ontologyVariableDataManager;
+
+	public void validate(final MeansImportRequest meansImportRequest) {
+
+		final BindingResult errors = new MapBindingResult(new HashMap<>(), MeansImportRequest.class.getName());
+
+		// Validate study
+		this.validateStudy(meansImportRequest);
+		// Check if means dataset already exists
+		this.checkIfMeansDatasetAlreadyExists(meansImportRequest);
```

Comment:
Let's extract study (id, instance, entries) validations to specific validator classes so they can be reused.

1. Checking of study ID- use  `StudyValidator.validate`
2. Checking of means dataset existence - use `StudyValidator.validateStudyHasNoMeansDataset`
3. Checking of environment numbers - Extract to `StudyValidator` as `validateStudyInstanceNumbers` (make the error code generic and not specific to means dataset)
4. Checking of study entries - Extract to `StudyEntryValidator` as `validateStudyContainsEntries(final Integer studyId, final List<Integer> entryNumbers)`